sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,824 @@
+        debug_assert!(index < self.len);
+        let offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+        &*(self.data.deref().as_ptr().add(offset) as *const SlotHashEntry)
+    }
+
+    /// Finds the hash for a specific slot using binary search.
+    ///
+    /// Returns the hash if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    #[inline(always)]
+    pub fn get_hash(&self, target_slot: Slot) -> Option<&[u8; HASH_BYTES]> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+            .map(|index| &entries[index].hash)
+    }
+
+    /// Finds the position (index) of a specific slot using binary search.
+    ///
+    /// Returns the index if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    #[inline(always)]
+    pub fn position(&self, target_slot: Slot) -> Option<usize> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (either the safe path that called `parse_and_validate_data` or
+    /// the unsafe `new_unchecked`) is responsible for ensuring the slice is big enough
+    /// and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        if self.len == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.data.deref().len() >= NUM_ENTRIES_SIZE + self.len * ENTRY_SIZE);
+
+        let entries_ptr =
+            unsafe { self.data.deref().as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry };
+        unsafe { core::slice::from_raw_parts(entries_ptr, self.len) }
+    }
+}
+
+impl<'a, T: Deref<Target = [u8]>> IntoIterator for &'a SlotHashes<T> {
+    type Item = &'a SlotHashEntry;
+    type IntoIter = core::slice::Iter<'a, SlotHashEntry>;
+
+    #[inline(always)]
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_entries_slice().iter()
+    }
+}
+
+impl<'a> SlotHashes<Ref<'a, [u8]>> {
+    /// Creates a `SlotHashes` instance by safely borrowing data from an `AccountInfo`.
+    ///
+    /// This function verifies that:
+    /// - The account key matches the `SLOTHASHES_ID`
+    /// - The data contains a valid length prefix
+    /// - The data is sufficiently large to hold the indicated number of entries
+    #[inline(always)]
+    pub fn from_account_info(account_info: &'a AccountInfo) -> Result<Self, ProgramError> {
+        if account_info.key() != &SLOTHASHES_ID {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        let data_ref = account_info.try_borrow_data()?;
+
+        let num_entries = Self::parse_and_validate_data(&data_ref)?;
```

Comment:
Perhaps the call to `parse_and_validate_data` should be part of a `new` constructor that does the checks.

When you are loading the data from the account, you can avoid this check as well as long as the `SLOTHASHES_ID` matches. In this way, the `from_account_info` method can "safely" use the `new_unchecked` constructor once you validate that it is the correct account.

When you are loading the data from an arbitrary byte slice, then you should use `new`, which will include all the necessary checks.